### PR TITLE
[FIX] website: prevent mega menu removal

### PR DESCRIPTION
**Pull request by smottet**
The user was able to remove the mega menu snippet without removing
the corresponding link in the menu. To avoid this flow to happen
we removed the delete button for the mega menu snippet. We also
handle the case where the user remove the mega menu snippet columns
one by one. On last column removal we put back the original template.

task-2636545

Description of the issue/feature this PR addresses:

Current behavior before PR:

Desired behavior after PR is merged:

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/website/static/src/js/editor/snippets.options.js [hunk 1]
```diff
@@ -3215,11 +3231,53 @@ options.registry.MegaMenuLayout = options.registry.SelectTemplate.extend({
...
+ * Resets Mega Menu template on last column deletion.
+ */
+options.registry.MegaMenuResetOnDelete = options.Class.extend({
+    /**
+     * @override
+     */
+    onRemove() {
+        this._super.apply(this, arguments);
+        if (this.$target.siblings('div[class^="col-"]').length === 0) {
+            core.bus.trigger('mega_menu_reset_template');
```
Comment (smottet) [inline addons/website/static/src/js/editor/snippets.options.js:3278]: Maybe we should do here what is done in `_onResetTemplate` ? It would avoid the `mega_menu_reset_template` event addition

File: addons/website/static/src/js/editor/snippets.options.js [hunk 2]
```diff
@@ -3205,6 +3205,22 @@ options.registry.MegaMenuLayout = options.registry.SelectTemplate.extend({
...
+     */
+    start() {
+        this._super(...arguments);
+        core.bus.on('mega_menu_reset_template', this, this._onResetTemplate);
+    },
+
+    /**
+     * @override
+     */
+     destroy: function () {
```
Comment (qsm-odoo) [inline addons/website/static/src/js/editor/snippets.options.js:3219]: Wrongly indented

File: addons/website/static/src/js/editor/snippets.options.js [hunk 3]
```diff
@@ -3205,6 +3205,22 @@ options.registry.MegaMenuLayout = options.registry.SelectTemplate.extend({
         this.selectTemplateWidgetName = 'mega_menu_template_opt';
     },
 
+    /**
+     * @override
+     */
+    start() {
+        this._super(...arguments);
```
Comment (qsm-odoo) [inline addons/website/static/src/js/editor/snippets.options.js:3212]: start is async, calling its super should take that into account

File: addons/website/static/src/js/editor/snippets.options.js [hunk 4]
```diff
@@ -3215,11 +3231,53 @@ options.registry.MegaMenuLayout = options.registry.SelectTemplate.extend({
...
+     */
+    async _onResetTemplate(ev) {
+        const widgetValue = this._computeWidgetState('selectTemplate');
+        const template = await this._rpc({
+            model: 'ir.ui.view',
+            method: 'render_public_asset',
+            args: [widgetValue, {}],
+        });
+        this.containerEl.insertAdjacentHTML('beforeend', template);
+    },
```
Comment (qsm-odoo) [inline addons/website/static/src/js/editor/snippets.options.js:3258]: Try to avoid remaking the RPC if not necessary, the SelectTemplate option has a cache.

File: addons/website/static/src/js/editor/snippets.options.js [hunk 5]
```diff
@@ -3215,11 +3231,53 @@ options.registry.MegaMenuLayout = options.registry.SelectTemplate.extend({
...
+});
+
+/**
+ * Hides delete button.
+ */
+options.registry.NoDeleteButton = options.Class.extend({
+    forceNoDeleteButton: true,
```
Comment (qsm-odoo) [inline addons/website/static/src/js/editor/snippets.options.js:3265]: Why not setting this on the already existing option targeting the mega menu?

File: addons/website/views/snippets/s_mega_menu_big_icons_subtitles.xml [hunk 6]
```diff
@@ -2,7 +2,7 @@
 <odoo>
 
 <template id="s_mega_menu_big_icons_subtitles" name="Menu - Big icons &amp; subtitles" groups="base.group_user">
-    <section class="s_mega_menu_big_icons_subtitles pt24 pb24" data-oe-shape-data="{'shape':'web_editor/Zigs/03','flip':[]}">
+    <section class="s_mega_menu s_mega_menu_big_icons_subtitles pt24 pb24" data-oe-shape-data="{'shape':'web_editor/Zigs/03','flip':[]}">
```
Comment (qsm-odoo) [inline addons/website/views/snippets/s_mega_menu_big_icons_subtitles.xml:5]: We should not need this. Besides, our convention for s_ classes is that `s_xxx_yyy` is either:
- A class put on a snippet root to uniquely identify it (like `s_mega_menu_big_icons_subtitles` here)
- A class put on a snippet's child which has an ancestor with the class `s_xxx`

File: addons/website/views/snippets/snippets.xml [hunk 7]
```diff
@@ -1004,24 +1004,27 @@
...
-                    <we-button data-select-label="Images Subtitles" title="Images Subtitles" data-select-template="website.s_mega_menu_images_subtitles" data-img="/website/static/src/img/snippets_thumbs/s_mega_menu_images_subtitles.svg">Images Subtitles</we-button>
-                    <we-button data-select-label="Logos" title="Logos" data-select-template="website.s_mega_menu_menus_logos" data-img="/website/static/src/img/snippets_thumbs/s_mega_menu_menus_logos.svg">Logos</we-button>
-                    <we-button data-select-label="Thumbnails" title="Thumbnails" data-select-template="website.s_mega_menu_thumbnails" data-img="/website/static/src/img/snippets_thumbs/s_mega_menu_thumbnails.svg">Thumbnails</we-button>
-                    <we-button data-select-label="Cards" title="Cards" data-select-template="website.s_mega_menu_cards" data-img="/website/static/src/img/snippets_thumbs/s_mega_menu_cards.svg">Cards</we-button>
-            </we-select>
+        <we-select string="Template" data-name="mega_menu_template_opt">
```
Comment (qsm-odoo) [inline addons/website/views/snippets/snippets.xml:1007]: Whoops, good fix but do that in a dedicated commit mentioning the commit which made the mistake.

File: addons/website/views/snippets/snippets.xml [hunk 8]
```diff
@@ -1004,24 +1004,27 @@
...
+            <we-button data-select-label="Thumbnails" title="Thumbnails" data-select-template="website.s_mega_menu_thumbnails" data-img="/website/static/src/img/snippets_thumbs/s_mega_menu_thumbnails.svg">Thumbnails</we-button>
+            <we-button data-select-label="Cards" title="Cards" data-select-template="website.s_mega_menu_cards" data-img="/website/static/src/img/snippets_thumbs/s_mega_menu_cards.svg">Cards</we-button>
+        </we-select>
         <we-select string="Size">
             <we-button data-select-class="">Full-Width</we-button>
             <we-button data-select-class="o_mega_menu_container_size">Narrow</we-button>
         </we-select>
     </div>
 
+    <div data-js="MegaMenuResetOnDelete" data-selector=".s_mega_menu .row > div " />
```
Comment (qsm-odoo) [inline addons/website/views/snippets/snippets.xml:1024]: It's not generic enough, it's too specific targeting columns. Instead, you can target `.o_mega_menu > *` which handles all cases and is very explicit. And you would avoid ugly code like "if there is not a column next to me after remove".

File: addons/website/static/src/js/editor/snippets.options.js (hunk 1 above)
Comment (qsm-odoo) [reply addons/website/static/src/js/editor/snippets.options.js:3278]: There is two problems with this:
- Using events on the bus is indeed ugly
- It does not account for the async part of the template being reset, source of race conditions.

I think this is possible to code using the `notify` method and the related `option_update` Odoo event. You can `trigger_up` that event, asking for the `MegaMenuLayout` option to do something and account for the async part returned.

**Review by qsm-odoo — CHANGES_REQUESTED** [review]

File: addons/website/static/src/js/editor/snippets.options.js (hunk 5 above)
Comment (smottet) [reply addons/website/static/src/js/editor/snippets.options.js:3265]: The existing option targets the Mega Menu and not the block below we actually want to target. 
megamenuselector

File: addons/website/static/src/js/editor/snippets.options.js (hunk 4 above)
Comment (smottet) [reply addons/website/static/src/js/editor/snippets.options.js:3258]: This cache is available only when the dropdown has been opened by the user. But yes I can check th ecache before making a RPC

File: addons/website/static/src/js/editor/snippets.options.js (hunk 5 above)
Comment (qsm-odoo) [reply addons/website/static/src/js/editor/snippets.options.js:3265]: Yep I saw that, forgot to remove that comment before sending the review. However, do not create an option named `NoDeleteButton`, that kinda makes wondering why there is a forceNoDeleteButton option per option if there is a dedicated Option Class to do that. Name is specifically something like "MegaMenuNoDelete".

File: addons/website/static/src/js/editor/snippets.options.js (hunk 4 above)
Comment (qsm-odoo) [reply addons/website/static/src/js/editor/snippets.options.js:3258]: My point goes both ways: you should also not update the cache on widget opening if the rpc was already done on template reset. And there is no need of code to handle that: you will **reduce** the code by implementing it properly:
- Dedicated function to do the RPC and update the cache
- Call that function on widget opening and on template reset

=> No duplicated code, les code and more efficient.

Comment (smottet) [reply addons/website/static/src/js/editor/snippets.options.js:3258]: Ok I see! Can't we also get rid of `_onWidgetOpening` and retrieve the template (from cache or through rpc) directly in selectTemplate ? 
Screenshot from 2021-09-14 09-45-04
Screenshot from 2021-09-14 09-44-56

File: addons/website/static/src/js/editor/snippets.options.js (hunk 1 above)
Comment (smottet) [reply addons/website/static/src/js/editor/snippets.options.js]: Not sure about what you mean by "_and account for the async part returned_"

File: addons/website/static/src/js/editor/snippets.options.js (hunk 4 above)
Comment (qsm-odoo) [reply addons/website/static/src/js/editor/snippets.options.js]: If you do that, you will have to wait for a RPC each first time you hover an option in that select which will thus create a flickering.

File: addons/website/static/src/js/editor/snippets.options.js (hunk 1 above)
Comment (qsm-odoo) [reply addons/website/static/src/js/editor/snippets.options.js]: Tenir compte de la promesse renvoyée, pas avoir une méthode qui lance des trucs async mais qui n'en tient pas compte.

File: addons/website/views/snippets/snippets.xml [hunk 9]
```diff
@@ -1004,24 +1004,27 @@
...
+            <we-button data-select-label="Thumbnails" title="Thumbnails" data-select-template="website.s_mega_menu_thumbnails" data-img="/website/static/src/img/snippets_thumbs/s_mega_menu_thumbnails.svg">Thumbnails</we-button>
+            <we-button data-select-label="Cards" title="Cards" data-select-template="website.s_mega_menu_cards" data-img="/website/static/src/img/snippets_thumbs/s_mega_menu_cards.svg">Cards</we-button>
+        </we-select>
         <we-select string="Size">
             <we-button data-select-class="">Full-Width</we-button>
             <we-button data-select-class="o_mega_menu_container_size">Narrow</we-button>
         </we-select>
     </div>
 
+    <div data-js="MegaMenuResetOnDelete" data-selector=".o_mega_menu > section > * " />
```
Comment (smottet) [inline addons/website/views/snippets/snippets.xml:1024]: @qsm-odoo We have this empty customize-block option now. What should I do about it? Hide it (I don't see straightforward way to do that)? Or let it as it is for potential future options.
Screenshot from 2021-09-15 14-09-42

File: addons/website/static/src/js/editor/snippets.options.js (hunk 1 above)
Comment (smottet) [reply addons/website/static/src/js/editor/snippets.options.js]: So the problem here is that onRemove will terminate before the action triggered ? 
https://github.com/odoo/odoo/pull/76399/files#diff-70f7fe38208aa7fe678f18e329d3c11b70065dee723921352b6005774e8bab53R3266-R3274

File: addons/website/views/snippets/snippets.xml (hunk 9 above)
Comment (qsm-odoo) [reply addons/website/views/snippets/snippets.xml:1024]: Not sure where it comes from but yes, find a way to hide it of course (and do that in a dedicated commit if it is relevant)

File: addons/website/static/src/js/editor/snippets.options.js (hunk 1 above)
Comment (qsm-odoo) [reply addons/website/static/src/js/editor/snippets.options.js]: Yes. Send a `onSuccess` callback with the option_update event so that it can be called once it is handled and onRemove being an async method, you can wait for that onSuccess call properly.

File: addons/website/views/snippets/snippets.xml (hunk 9 above)
Comment (smottet) [reply addons/website/views/snippets/snippets.xml:1024]: It is because the selector doesn't target the row anymore but the container just above. Should customize-block option without any visible option inside always be hidden? I imagine this is not the only case where we have such empty customize-block. So may be it is worth doing something generic for this case.
Screenshot from 2021-09-15 14-35-04

Comment (qsm-odoo) [reply addons/website/views/snippets/snippets.xml:1024]: Why is there an option targeting the container, this is what I don't understand

Comment (smottet) [reply addons/website/views/snippets/snippets.xml:1024]: If the user removes manually all the cols inside it then the container will be removed and the onRemove called. We can't target the cols and reset the template if no neighbors due to this kind of template. The data-selector here is generic enough for all the mega_menu templates...
Screenshot from 2021-09-15 14-54-44

Comment (qsm-odoo) [reply addons/website/views/snippets/snippets.xml:1024]: Why not targeting the section itself?

Comment (smottet) [reply addons/website/views/snippets/snippets.xml:1024]: oops indeed

**Comment by smottet** [comment]
@qsm-odoo  I made the changes

**Review by qsm-odoo — CHANGES_REQUESTED** [review]
Seems good :+1: Just some details.

**Review by qsm-odoo — COMMENTED** [review]
@smottet I push-forced the branch:
- https://github.com/odoo/odoo/compare/1d7d475891cb6fa3ceb278b2132928fe5638a1a1..32da13958647dd05d21e68dcf6eda566b24a6ebd
    - Failed space in front of `notify` definition
    - `notify` cannot be `async`, so I adapted accordingly
    - Another indentation fail
- I though `onRemove` could already be async but it was not the case: I added a commit to properly allow it: https://github.com/odoo/odoo/pull/76399/commits/b583627ec775a7a54d8f9be09d841038d3a321ae
- Doing that, I found that `onRemove`/`onClone` editor creations (and now `async` `onRemove` calls) were not properly awaited, I fixed it here but will merge that in stable and wait for the forward-port: https://github.com/odoo/odoo/pull/76399/commits/055e827b1969b5106e1f1c5dd6872bed1d1d9f46

**Comment by smottet** [comment]
@qsm-odoo Ok I'll have a look at these changes

**Review by qsm-odoo — APPROVED** [review]
@robodoo rebase-ff r+

**Comment by qsm-odoo** [comment]
@robodoo rebase-ff r+
